src/plugins/presentation_util/public/components/floating_actions/floating_actions.tsx

```diff
@@ -41,46 +46,68 @@ export const FloatingActions: FC<FloatingActionsProps> = ({
   className = '',
   disabledActions,
 }) => {
-  const [floatingActions, setFloatingActions] = useState<JSX.Element | undefined>(undefined);
+  const [floatingActions, setFloatingActions] = useState<FloatingActionItem[]>([]);
 
   useEffect(() => {
     if (!api) return;
 
-    const getActions = async () => {
-      let mounted = true;
-      const context = {
-        embeddable: api,
-        trigger: panelHoverTrigger,
-      };
+    let mounted = true;
+    const context = {
+      embeddable: api,
+      trigger: panelHoverTrigger,
+    };
+
+    const getActions: () => Promise<FloatingActionItem[]> = async () => {
       const actions = (
         await uiActionsService.getTriggerCompatibleActions(PANEL_HOVER_TRIGGER, context)
       )
-        .filter((action): action is Action & { MenuItem: React.FC<{ context: unknown }> } => {
+        .filter((action) => {
           return action.MenuItem !== undefined && (disabledActions ?? []).indexOf(action.id) === -1;
         })
         .sort((a, b) => (a.order || 0) - (b.order || 0));
+      return actions as FloatingActionItem[];
+    };
+
+    const subscriptions = new Subscription();
 
+    const handleActionCompatibilityChange = (isCompatible: boolean, action: Action) => {
       if (!mounted) return;
-      if (actions.length > 0) {
-        setFloatingActions(
-          <>
-            {actions.map((action) =>
-              React.createElement(action.MenuItem, {
-                key: action.id,
-                context,
-              })
-            )}
-          </>
+      setFloatingActions((currentActions) => {
+        const newActions = currentActions
+          ?.filter((current) => current.id !== action.id)
+          .sort((a, b) => (a.order || 0) - (b.order || 0));
```

Comment:
Could we maybe make the filter + sort a helper method? Since it's used both here and in `getActions`

Reply:
I was playing around and came up with this way, but maybe there is another way I haven't thought of...
Tell me what you think 😊
So the `filter` functions seem to be different, and only `sort` is the same.
So the potential code snippet might look like this:

```
   const sortByOrder = (array: Action[]) => {
      return array.sort((a, b) => (a.order || 0) - (b.order || 0));
    };

    const getActions: () => Promise<FloatingActionItem[]> = async () => {
      return sortByOrder(
        await uiActionsService.getTriggerCompatibleActions(PANEL_HOVER_TRIGGER, context)
      ).filter((action: Action) => {
        return action.MenuItem !== undefined && (disabledActions ?? []).indexOf(action.id) === -1;
      }) as FloatingActionItem[];
    };
```
Do you think it is a better way than what is currently in the code? Or did you have something else in mind?

Reply:
Separating out just the `sort` looks good to me! :) But it can actually be done a little more efficiently:
```typescript
   const sortByOrder = (a: Action, b: Action) => {
      return (a.order || 0) - (b.order || 0);
    };

    const getActions: () => Promise<FloatingActionItem[]> = async () => {
      return (
        await uiActionsService.getTriggerCompatibleActions(PANEL_HOVER_TRIGGER, context)
      )
        .filter((action) => {
          return action.MenuItem !== undefined && (disabledActions ?? []).indexOf(action.id) === -1;
        })
        .sort(sortByOrder) as FloatingActionItem[];
    };
```

Sorting is a much more expensive operation than filtering, especially for larger arrays - so it's good practice to filter **first** before sorting :) Obviously it's not super crucial in this instance, since the floating actions is never more than 3 items, but I still think this code is a little easier to read. What do you think?

This is very much a nit a this point, so I'm good with your solution, too 🤷

Reply:
Wow! Thank you for this explanation 🙏🏻
✅